src/backend/distributed/connection/connection_management.c

```diff
@@ -402,6 +402,40 @@ CloseConnectionByPGconn(PGconn *pqConn)
 
 
 /*
+ * ShutdownConnection, if necessary cancels the currently running statement,
+ * and then closes the underlying libpq connection.  The MultiConnection
+ * itself is left intact.
+ *
+ * NB: Cancelling a statement requires network IO, and currently is not
+ * interruptible. Unfortunately libpq does not provide a non-blocking
+ * implementation of PQcancel(), so we don't have much choice for now.
```

Comment:
Yea, it really sucks.  I'm going to try to get a better implementation of this into postgres, but it'll be a while before we can use it.  It's a very different code-path from normal libpq, because it needs to be able to run from a signal handler, so it might be suitable to be copied into our code at some point.